Fix backup dates

JoaoJandre opened the pull request:
### Description

Currently, the backup date in Veeam is collected as a `String` and saved as `VARCHAR(255)` in the database, which can lead to errors and inconsistencies.

We are proposing with this PR to persist the backup dates as a `DATETIME` object, which will enable us to properly manipulate it, regardless of the date format used in the environment. The UI was also changed to display the complete data format.

When upgrading the DB schema, we delete all backup entries in the DB; these backups are synchronized with Veeam backups by ACS afterwards. In the end, no Veeam backups should be lost. Only dummy backups will be permanently deleted, and since they are not relevant in production, no harm will be caused.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

### Feature/Enhancement Scale or Bug Severity

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial


### Screenshots (if appropriate):

Storage->Backups
![image](https://user-images.githubusercontent.com/48719461/174315752-68bb7273-84d9-414d-b742-76611f6315be.png)

Instances->VM->Backups
![image](https://user-images.githubusercontent.com/48719461/174315868-6f5e7606-2d39-4cc1-972b-4bc35257d5a6.png)


### How Has This Been Tested?
This was tested in a local lab, by upgrading the DB schema and then checking if ACS would correctly synchronize with Veeam. After that, the UI was checked to see if everything was as expected. 
New backups were created and checked. The same tests were made using the dummy provider.

blueorangutan commented:
@GutoVeronezi a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

GutoVeronezi commented on engine/schema/src/main/resources/META-INF/db/schema-41710to41800.sql:
```diff
@@ -17,4 +17,11 @@
+WHERE 1;
+
+ALTER TABLE `cloud`.`backups`
+MODIFY COLUMN `date` DATETIME;
```
We need to add a newline at the end of the file.

blueorangutan commented:
UI build: :heavy_multiplication_x:
 (SL-JID-1799)

blueorangutan commented:
UI build: :heavy_multiplication_x:
 (SL-JID-3)

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 1

blueorangutan commented:
Packaging result: :heavy_multiplication_x: el7 :heavy_multiplication_x: el8 :heavy_multiplication_x: debian :heavy_multiplication_x: suse15. SL-JID 3617

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_multiplication_x:
 (SL-JID-1810)

DaanHoogland commented on plugins/backup/veeam/src/main/java/org/apache/cloudstack/backup/veeam/VeeamClient.java:
```diff
@@ -636,15 +638,19 @@ public Map<String, Backup.Metric> getBackupMetrics() {
                 String [] split = part.split(":", 2);
-                created = split[1].trim();
+                split[1] = StringUtils.trim(split[1]);
+                String [] time = split[1].split(":|/| ");
```
```suggestion
                String [] time = split[1].split("[:/ ]");
```

blueorangutan commented:
@acs-robot a Jenkins job has been kicked to build UI QA env. I'll keep you posted as I make progress.

blueorangutan commented:
UI build: :heavy_multiplication_x:
 (SL-JID-1837)

DaanHoogland reviewed (approved):
CLGTM

JoaoJandre commented:
> @JoaoJandre which versions of Veeam was this tested with?

Hello @rohityadavcloud, no changes were made in the way the data is being collected. Only in the way the data is processed and stored in the DB. Therefore, this will work with the same versions of Veeam as it has been working until now.

rohityadavcloud commented on engine/schema/src/main/resources/META-INF/db/schema-41710to41800.sql:
```diff
@@ -22,4 +22,11 @@
+WHERE so.default_use = 1 AND so.vm_type IN ('domainrouter', 'secondarystoragevm', 'consoleproxy', 'internalloadbalancervm', 'elasticloadbalancervm');
+
+-- Drop all backup records and change date column type to DATETIME. The data will be resynchronized automatically later;
+DELETE FROM `cloud`.`backups`
```
Wouldn't that cause data loss for existing backup entries? Why not modify the column date/time?

harikrishna-patnala replied:
@JoaoJandre I see your point that even if these backups are deleted from the database, the management server will again sync and get all the backups from Veeam. But instead of losing the history of backups, it is better to modify the date. You can use java upgrade file Upgrade417100to418000.java file to do that.

By deleting the entries we will also lose the history of removed backups.

sureshanaparti reviewed (approved):
code LGTM, needs testing.

@JoaoJandre make sure the upgrade for backup data in DB is properly handled (without any data loss).

RodrigoDLopez reviewed (approved):
lgtm

rohityadavcloud commented:
This PR breaks compatibility for existing users with backups on upgrade as the upgrade path runs ```DELETE FROM `cloud`.`backups` ```. So, it can't be accepted unless the PR handles upgrades in a more smoother way, or at least we make an attempt to consult and build support with users (on users@ perhaps?).

DaanHoogland commented:
@JoaoJandre can you look at the conflicts here, please?

JoaoJandre commented:
@rohityadavcloud @harikrishna-patnala @sureshanaparti I finally got around addressing your reviews, now the data migration is done in the Upgrade41810to41900.java file.

DaanHoogland commented on engine/schema/src/main/resources/META-INF/db/schema-41720to41800-cleanup.sql:
```diff
@@ -16,5 +16,5 @@
 
 --;
--- Schema upgrade cleanup from 4.17.2.0 to 4.18.0.0
+-- Schema upgrade from 4.17.1.0 to 4.18.0.0
```
is this an accident?

JoaoJandre replied:
Yes, i'll fix it

rohityadavcloud commented:
@blueorangutan package

blueorangutan commented:
@rohityadavcloud a [SF] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 6106

DaanHoogland commented:
@blueorangutan test rocky-8 vmware-72

blueorangutan commented:
@DaanHoogland [SF] unsupported parameters provided. Supported mgmt server os are: `centos7, centos6, suse15, alma8, ubuntu18, ubuntu22, ubuntu20, rocky8, alma9`. Supported hypervisors are: `kvm-centos6, kvm-centos7, kvm-rocky8, kvm-alma8, kvm-alma9, kvm-ubuntu18, kvm-ubuntu20, kvm-ubuntu22, kvm-suse15, vmware-55u3, vmware-60u2, vmware-65u2, vmware-67u3, vmware-70u1, vmware-70u2, vmware-70u3, vmware-80, vmware-80u1, xenserver-65sp1, xenserver-71, xenserver-74, xcpng74, xcpng76, xcpng80, xcpng81, xcpng82`

DaanHoogland commented:
@blueorangutan test rocky8 vmware-70u3

blueorangutan commented:
@DaanHoogland a [SF] Trillian-Jenkins test job (rocky8 mgmt + vmware-70u3) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-6573)</b>
Environment: vmware-70u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 56058 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6473-t6573-vmware-70u3.zip
Smoke tests completed. 108 look OK, 2 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_deploy_vm_on_specific_host | `Error` | 3602.55 | test_vm_deployment_planner.py
test_02_deploy_vm_on_specific_cluster | `Error` | 4.35 | test_vm_deployment_planner.py
test_03_deploy_vm_on_specific_pod | `Error` | 4.37 | test_vm_deployment_planner.py
test_04_deploy_vm_on_host_override_pod_and_cluster | `Error` | 2.32 | test_vm_deployment_planner.py
test_05_deploy_vm_on_cluster_override_pod | `Error` | 2.33 | test_vm_deployment_planner.py
test_09_expunge_vm | `Failure` | 424.61 | test_vm_life_cycle.py

DaanHoogland commented:
@blueorangutan LLtest rocky8 vmware-70u3

blueorangutan commented:
@DaanHoogland a [LL] Trillian-Jenkins test job (rocky8 mgmt + vmware-70u3) has been kicked to run smoke tests

JoaoJandre commented:
@DaanHoogland could we run the CI again for this PR as well?

DaanHoogland commented:
@blueorangutan package

blueorangutan commented:
@DaanHoogland a [SF] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 6475

DaanHoogland commented:
@blueorangutan test rocky8 vmware-67u3

blueorangutan commented:
@DaanHoogland a [SF] Trillian-Jenkins test job (rocky8 mgmt + vmware-67u3) has been kicked to run smoke tests

harikrishna-patnala commented:
changed the label to 4.19 as the upgrade file changes are related to 4.19 version

JoaoJandre commented:
@DaanHoogland I think BO ended up not running the tests on this one, can we retry?

JoaoJandre commented:
@blueorangutan package

blueorangutan commented:
@JoaoJandre a [SF] Jenkins job has been kicked to build packages. It will be bundled with  KVM, XenServer and VMware SystemVM templates. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result [SF]: :heavy_check_mark: el7 :heavy_check_mark: el8 :heavy_check_mark: el9 :heavy_check_mark: debian :heavy_check_mark: suse15. SL-JID 6723

DaanHoogland commented:
@blueorangutan test rocky8 vmware-67u3

blueorangutan commented:
@DaanHoogland a [SF] Trillian-Jenkins test job (rocky8 mgmt + vmware-67u3) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-7358)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 83324 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6473-t7358-vmware-67u3.zip
Smoke tests completed. 86 look OK, 27 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_list_cpvm_vm | `Failure` | 0.04 | test_ssvm.py
test_04_cpvm_internals | `Failure` | 0.04 | test_ssvm.py
ContextSuite context=TestHostControlState>:setup | `Error` | 4.50 | test_host_control_state.py
test_01_add_delete_kubernetes_supported_version | `Error` | 0.07 | test_kubernetes_supported_versions.py
ContextSuite context=TestListIdsParams>:setup | `Error` | 0.00 | test_list_ids_parameter.py
ContextSuite context=TestRouterDns>:setup | `Error` | 0.00 | test_router_dns.py
ContextSuite context=TestLoadBalance>:setup | `Error` | 0.00 | test_loadbalance.py
ContextSuite context=TestRouterDnsService>:setup | `Error` | 0.00 | test_router_dnsservice.py
ContextSuite context=TestRouterServices>:setup | `Error` | 0.00 | test_routers.py
ContextSuite context=TestMetrics>:setup | `Error` | 0.00 | test_metrics_api.py
ContextSuite context=TestIsolatedNetworks>:setup | `Error` | 0.00 | test_routers_network_ops.py
ContextSuite context=TestRedundantIsolateNetworks>:setup | `Error` | 0.00 | test_routers_network_ops.py
ContextSuite context=TestRouterIpTablesPolicies>:setup | `Error` | 0.00 | test_routers_iptables_default_policy.py
ContextSuite context=TestVPCIpTablesPolicies>:setup | `Error` | 0.00 | test_routers_iptables_default_policy.py
test_01_sys_vm_start | `Failure` | 60.21 | test_secondary_storage.py
ContextSuite context=TestServiceOfferings>:setup | `Error` | 1516.18 | test_service_offerings.py
ContextSuite context=TestVMWareStoragePolicies>:setup | `Error` | 0.00 | test_storage_policy.py
test_02_create_template_with_checksum_sha1 | `Error` | 65.58 | test_templates.py
test_03_create_template_with_checksum_sha256 | `Error` | 65.59 | test_templates.py
test_04_create_template_with_checksum_md5 | `Error` | 65.58 | test_templates.py
test_05_create_template_with_no_checksum | `Error` | 65.59 | test_templates.py
ContextSuite context=TestTemplates>:setup | `Error` | 6.76 | test_templates.py
ContextSuite context=TestVmAutoScaling>:setup | `Error` | 0.00 | test_vm_autoscaling.py
test_01_deploy_vm_on_specific_host | `Error` | 1.32 | test_vm_deployment_planner.py
test_02_deploy_vm_on_specific_cluster | `Error` | 1.27 | test_vm_deployment_planner.py
test_03_deploy_vm_on_specific_pod | `Error` | 1.31 | test_vm_deployment_planner.py
test_04_deploy_vm_on_host_override_pod_and_cluster | `Error` | 1.34 | test_vm_deployment_planner.py
test_05_deploy_vm_on_cluster_override_pod | `Error` | 1.32 | test_vm_deployment_planner.py
ContextSuite context=TestDeployVM>:setup | `Error` | 0.00 | test_vm_life_cycle.py
test_01_offline_migrate_VM_and_root_volume | `Error` | 1.33 | test_vm_life_cycle.py
test_02_offline_migrate_VM_with_two_data_disks | `Error` | 1.30 | test_vm_life_cycle.py
test_03_live_migrate_VM_with_two_data_disks | `Error` | 1.29 | test_vm_life_cycle.py
test_04_migrate_detached_volume | `Error` | 1.32 | test_vm_life_cycle.py
ContextSuite context=TestVAppsVM>:setup | `Error` | 1518.65 | test_vm_life_cycle.py
ContextSuite context=TestVMLifeCycle>:setup | `Error` | 1522.76 | test_vm_life_cycle.py
test_01_unmanage_vm_cycle | `Error` | 4.95 | test_vm_lifecycle_unmanage_import.py
ContextSuite context=TestUnmanageVM>:teardown | `Error` | 4.97 | test_vm_lifecycle_unmanage_import.py
ContextSuite context=TestVMSchedule>:setup | `Error` | 0.00 | test_vm_schedule.py
test_change_service_offering_for_vm_with_snapshots | `Error` | 1.35 | test_vm_snapshots.py
ContextSuite context=TestVmSnapshot>:setup | `Error` | 10.57 | test_vm_snapshots.py
ContextSuite context=TestCreateVolume>:setup | `Error` | 0.00 | test_volumes.py
ContextSuite context=TestVolumes>:setup | `Error` | 1516.36 | test_volumes.py
ContextSuite context=TestIpv6Vpc>:setup | `Error` | 0.00 | test_vpc_ipv6.py
ContextSuite context=TestVPCRedundancy>:setup | `Error` | 0.00 | test_vpc_redundant.py
ContextSuite context=TestVPCNics>:setup | `Error` | 0.00 | test_vpc_router_nics.py
ContextSuite context=TestRVPCSite2SiteVpn>:setup | `Error` | 0.00 | test_vpc_vpn.py
ContextSuite context=TestVPCSite2SiteVPNMultipleOptions>:setup | `Error` | 0.00 | test_vpc_vpn.py
ContextSuite context=TestVpcRemoteAccessVpn>:setup | `Error` | 0.00 | test_vpc_vpn.py
ContextSuite context=TestVpcSite2SiteVpn>:setup | `Error` | 0.00 | test_vpc_vpn.py
test_02_cancel_host_maintenace_with_migration_jobs | `Error` | 1.63 | test_host_maintenance.py
test_03_cancel_host_maintenace_with_migration_jobs_failure | `Error` | 1.67 | test_host_maintenance.py

JoaoJandre commented:
I looked at some of these errors and none of them seemed related to this PR, could we rerun the tests?

DaanHoogland commented:
@blueorangutan test rocky8 vmware-67u3

blueorangutan commented:
@DaanHoogland a [SF] Trillian-Jenkins test job (rocky8 mgmt + vmware-67u3) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-7378)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 72244 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6473-t7378-vmware-67u3.zip
Smoke tests completed. 109 look OK, 4 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_trigger_shutdown | `Failure` | 337.13 | test_safe_shutdown.py
test_list_vms_metrics_admin | `Error` | 3614.90 | test_metrics_api.py
test_list_vms_metrics_history | `Error` | 5.67 | test_metrics_api.py
test_list_volumes_metrics_history | `Error` | 3623.34 | test_metrics_api.py
test_01_deploy_vm_on_specific_host | `Error` | 18.76 | test_vm_deployment_planner.py
test_02_deploy_vm_on_specific_cluster | `Error` | 3603.64 | test_vm_deployment_planner.py
test_03_deploy_vm_on_specific_pod | `Error` | 4.45 | test_vm_deployment_planner.py
test_04_deploy_vm_on_host_override_pod_and_cluster | `Error` | 4.43 | test_vm_deployment_planner.py
test_05_deploy_vm_on_cluster_override_pod | `Error` | 2.33 | test_vm_deployment_planner.py
test_09_expunge_vm | `Failure` | 426.69 | test_vm_life_cycle.py

blueorangutan commented:
<b>[SF] Trillian test result (tid-7386)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 58837 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6473-t7386-vmware-67u3.zip
Smoke tests completed. 112 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_invalid_upgrade_kubernetes_cluster | `Failure` | 3610.45 | test_kubernetes_clusters.py
test_02_upgrade_kubernetes_cluster | `Failure` | 3619.28 | test_kubernetes_clusters.py
test_03_deploy_and_scale_kubernetes_cluster | `Failure` | 0.05 | test_kubernetes_clusters.py
test_04_autoscale_kubernetes_cluster | `Failure` | 0.06 | test_kubernetes_clusters.py
test_05_basic_lifecycle_kubernetes_cluster | `Failure` | 0.06 | test_kubernetes_clusters.py
test_06_delete_kubernetes_cluster | `Failure` | 0.08 | test_kubernetes_clusters.py
test_07_deploy_kubernetes_ha_cluster | `Failure` | 0.08 | test_kubernetes_clusters.py
test_08_upgrade_kubernetes_ha_cluster | `Failure` | 0.05 | test_kubernetes_clusters.py
test_09_delete_kubernetes_ha_cluster | `Failure` | 0.06 | test_kubernetes_clusters.py
test_10_vpc_tier_kubernetes_cluster | `Failure` | 72.65 | test_kubernetes_clusters.py
test_11_test_unmanaged_cluster_lifecycle | `Error` | 1.32 | test_kubernetes_clusters.py
ContextSuite context=TestKubernetesCluster>:teardown | `Error` | 68.26 | test_kubernetes_clusters.py

JoaoJandre commented:
@blueorangutan package

DaanHoogland commented:
@blueorangutan test rocky8 vmware-67u3

blueorangutan commented:
@DaanHoogland a [SF] Trillian-Jenkins test job (rocky8 mgmt + vmware-67u3) has been kicked to run smoke tests

blueorangutan commented:
<b>[SF] Trillian test result (tid-7456)</b>
Environment: vmware-67u3 (x2), Advanced Networking with Mgmt server r8
Total time taken: 48802 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr6473-t7456-vmware-67u3.zip
Smoke tests completed. 112 look OK, 1 have errors, 0 did not run
Only failed and skipped tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_02_upgrade_kubernetes_cluster | `Failure` | 588.15 | test_kubernetes_clusters.py
test_04_autoscale_kubernetes_cluster | `Failure` | 71.02 | test_kubernetes_clusters.py
